extensions/src/main/java/org/pentaho/platform/web/http/api/resources/Log4jResource.java

```diff
@@ -0,0 +1,103 @@
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details.
+ *
+ * Copyright (c) 2017 Pentaho Corporation..  All rights reserved.
+ */
+
+package org.pentaho.platform.web.http.api.resources;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.log4j.Level;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.log4j.helpers.Loader;
+import org.apache.log4j.xml.DOMConfigurator;
+import org.codehaus.enunciate.jaxrs.ResponseCode;
+import org.codehaus.enunciate.jaxrs.StatusCodes;
+
+import javax.ws.rs.FormParam;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.Enumeration;
+
+
+/**
+ * Created by on 6/23/2017.
+ */
+@Path( "/logconfig" )
+public class Log4jResource {
+
+  private static final Logger LOGGER = Logger.getLogger( Log4jResource.class );
+  private static final String CONFIG = "log4j.xml";
+
+  @PUT
+  @Path ( "/reset" )
+  @StatusCodes( {
+    @ResponseCode( code = 200, condition = "Successfully reset log level." )
+    } )
+  @Produces( { MediaType.TEXT_PLAIN } )
+  public Response resetLogLevel() throws Exception {
+    LOGGER.setLevel( Level.INFO );
+    LOGGER.info( "Resetting Log Level..." );
+
+    DOMConfigurator.configure( Loader.getResource( CONFIG ) );
+    LOGGER.info( LogManager.getRootLogger().getLevel() );
+    return Response.ok( "Done" ).build();
+  }
+
+  @PUT
+  @Path ( "/update" )
+  @StatusCodes( {
+    @ResponseCode( code = 200, condition = "Successfully update log level." ),
+    @ResponseCode( code = 304, condition = "Log level is not modified." )
+    } )
+  @Produces( { MediaType.TEXT_PLAIN } )
+  public Response updateLogLevel( @FormParam(  "level" ) String targetLevel, @FormParam( "cat" ) String category ) throws Exception {
+    LOGGER.setLevel( Level.INFO );
+
+    if ( StringUtils.isBlank( targetLevel ) && StringUtils.isBlank( category ) ) {
+      return Response.notModified( "No parameter provided, log level not modified." ).build();
+    }
+
+    Logger root = LogManager.getRootLogger();
+
+    if ( StringUtils.isNotBlank( targetLevel ) ) {
+      LOGGER.info( "Request to set log level: " + targetLevel );
+
+      if ( StringUtils.isNotBlank( category ) ) {
+        LOGGER.info( "Request to set log level for package: " + category );
+        Logger catLog = LogManager.exists( category );
+        if ( catLog != null ) {
+          catLog.setLevel( Level.toLevel( targetLevel, root.getLevel() ) );
+          return Response.ok( "Setting log level for: '" + catLog.getName() + "' to be: " + catLog.getLevel() ).build();
+        }
+        return Response.notModified( "Category: '" + category + "' not found, log level not modified." ).build();
```

Comment:
is it possible to add a category if it doesn't exist? I think it would be immensely helpful. Currently, to see logs from the dispatching code, we have to add a number of categories that don't currently exist. We may want to just add them to log4j.xml and ship with the added category, but I think, if at all possible, It would go a long way to be able to add categories on the fly.